File: nvxapp.client/src/nvx/ClientServer-Service/GestionePresenze/Dip_GG_Timbratura/Models/dip-gg-timbratura-model.ts
import { ModelResult } from "../../../ModelsBase/model-result";
import { StatoRichiesta } from "../../Dip_GG_Richiesta/Models/dip-gg-richiesta-model";
import { Par_AttivitaModel } from "../../Par_Attivita/Models/par-attivita-model";



export class Dip_GG_TimbraturaModel {

  public id: number;
  public idDip_RapportoLavoro!: number;

  public timbratura!: Date;
  public timbraturaOriginale!: Date;
  public timbraturaArrotondata?: Date;
  public giornoCompetenza!: Date;

  public timbraturaTipo!: TipoTimbratura;

  public richiestaStato!: StatoRichiesta;
  public idDip_GG_Richiesta?: number;
  public idAz_SubCommessaAttivita: number;

}
export enum TipoTimbratura {
  Entrata,
  Uscita,
  SenzaVerso,
  Attivita
}



export class Dip_GG_Timbratura_GetAll_InModel {
  public idAspNetUsers?: string
  public year: number;
  public month: number;
}
export class Dip_GG_Timbratura_GetAll_OutModel extends ModelResult {
  public dip_GG_Timbratura: Dip_GG_TimbraturaModel[];
}


export class Dip_GG_Timbratura_Stamp_InModel {
  public dateStamp!: string;
  public excludeRicalc: boolean;
  public idAz_SubCommessaAttivita: number;
}
export class Dip_GG_Timbratura_Stamp_OutModel extends ModelResult {
}


export class Dip_GG_Timbratura_Get_4Calculation_InModel {
  public usersId: string[] = [];
  public dal!: string; // ISO 8601 locale, es. "2026-03-01T00:00:00"
  public al!: string;  // ISO 8601 locale, es. "2026-03-31T23:59:59"
}
export class Dip_GG_Timbratura_Get_4Calculation_OutModel extends ModelResult {
  public dip_GG_Timbratura: Dip_GG_TimbraturaModel[] = [];
}


export class Dip_GG_TimbraturaGetInModel {
  public id: number;

  public data?: string;
  public idDip_RapportoLavoro!: number;
}
export class Dip_GG_TimbraturaGetOutModel extends ModelResult {
  public dip_GG_Timbratura: Dip_GG_TimbraturaModel;
}


export class Dip_GG_TimbraturaPutInModel {
  public excludeRicalc: boolean;
  public dip_GG_Timbratura: Dip_GG_TimbraturaModel;
}
export class Dip_GG_TimbraturaPutOutModel extends ModelResult {
  public dip_GG_Timbratura: Dip_GG_TimbraturaModel;
}



export class Dip_GG_Timbratura_DeleteInModel {
  public excludeRicalc: boolean;
  public id: number;
}
export class Dip_GG_Timbratura_DeleteOutModel extends ModelResult {

}

export class Dip_GG_Timbratura_StampPrepare_InModel {
  public idAspNetUsers?: string;
}
export class Dip_GG_Timbratura_StampPrepare_OutModel extends ModelResult {
  public idAspNetUsers: string = '';
  public currentDate!: Date;
  public idPar_Attivita: number | null = null;
  public par_Attivita: Par_AttivitaModel[] = [];
}

File: nvxapp.server.service/ClientServer-Service/GestionePresenze/Dip_GG_TimbraturaService/Models/Dip_GG_TimbraturaModel.cs
using nvxapp.server.data.Entities.Tenant;
using nvxapp.server.service.ClientServer_Service.ModelsBase;

namespace nvxapp.server.service.ClientServer_Service.GestionePresenze.Dip_GG_TimbraturaService.Models
{
    public class Dip_GG_TimbraturaModel : HashModel
    {
        [HashField]
        public int Id { get; set; }

        [HashField]
        public int IdDip_RapportoLavoro { get; set; }

        [HashField]
        public DateTime Timbratura { get; set; }
        [HashField]
        public DateTime TimbraturaOriginale { get; set; }
        [HashField]
        public DateTime? TimbraturaArrotondata { get; set; }
        [HashField]
        public DateTime GiornoCompetenza { get; set; } // girno al quale viene agganciata la timbratura (servirà per cavallo notte montanti /smontanti)
        [HashField]
        public TipoTimbratura TimbraturaTipo { get; set; }

        /* 
          per gli inserimenti diretti, 
                StatoRichiasta = Diretta e
                IdDip_GG_Richiesta = null
         */
        [HashField]
        public StatoRichiesta RichiestaStato { get; set; }
        [HashField]
        public int? idDip_GG_Richiesta { get; set; }


    }



    public class Dip_GG_Timbratura_GetAll_InModel
    {
        public string? IdAspNetUsers { get; set; }
        public int Year { get; set; } = 0;
        public int Month { get; set; } = 0;
    }
    public class Dip_GG_Timbratura_GetAll_OutModel : ModelResult
    {
        public List<Dip_GG_TimbraturaModel> Dip_GG_Timbratura { get; set; } = new List<Dip_GG_TimbraturaModel>();

    }



    public class Dip_GG_Timbratura_Stamp_InModel
    {
        public string DateStamp { get; set; } = string.Empty;
    }
    public class Dip_GG_Timbratura_Stamp_OutModel : ModelResult
    {

    }

    public class Dip_GG_Timbratura_Get_4Calculation_InModel
    {
        public List<string> UsersId { get; set; } = new List<string>();
        public DateTime Dal { get; set; }
        public DateTime Al { get; set; }
    }

    public class Dip_GG_Timbratura_Get_4Calculation_OutModel : ModelResult
    {
        public List<Dip_GG_TimbraturaModel> Dip_GG_Timbratura { get; set; } = new List<Dip_GG_TimbraturaModel>();
    }

    public class Dip_GG_TimbraturaGetInModel
    {
        public int Id { get; set; }

        /* per inizializzare il record nuovo */
        public required int IdDip_RapportoLavoro { get; set; }
        public DateTime? Data { get; set; }
    }
    public class Dip_GG_TimbraturaGetOutModel : ModelResult
    {
        public Dip_GG_TimbraturaModel Dip_GG_Timbratura { get; set; } = new Dip_GG_TimbraturaModel();
    }
    
    public class Dip_GG_TimbraturaPutInModel
    {
        public int IdDip_RapportoLavoro { get; set; }
        public Dip_GG_TimbraturaModel Dip_GG_Timbratura { get; set; } = new Dip_GG_TimbraturaModel() { Id = 0, IdDip_RapportoLavoro = 0 };
    }
    public class Dip_GG_TimbraturaPutOutModel : ModelResult
    {
        public Dip_GG_TimbraturaModel Dip_GG_Timbratura { get; set; } = new Dip_GG_TimbraturaModel() { Id = 0, IdDip_RapportoLavoro = 0 };
    }

    public class Dip_GG_Timbratura_DeleteInModel
    {
        public int Id { get; set; }
    }
    public class Dip_GG_Timbratura_DeleteOutModel : ModelResult { }


}
